README.md
# Factotum Server

[![Release][release-image]][releases] [![License][license-image]][license]

## Overview

Job server as a web service. Enables the scheduling and concurrent execution of Factotum jobs.  See the [user guide][user-guide] for further information on getting up and running.

## User quickstart

Assuming you're running **64 bit Linux**:

```{bash}
$ wget https://github.com/snowplow/factotum-server/releases/download/0.3.0/factotum_server_0.3.0_linux_x86_64.zip
$ unzip factotum_server_0.3.0_linux_x86_64.zip
$ ./factotum-server --help
```

[Factotum][factotum-repo] needs to be available locally. You can then run Factotum Server with preset defaults and the Factotum path specified in the following way:

```{bash}
$ ./factotum-server --factotum-bin=<PATH>
```

If Factotum is **not** already available:

```{bash}
$ wget https://github.com/snowplow/factotum/releases/download/0.6.0/factotum_0.6.0_linux_x86_64.zip
$ unzip factotum_0.6.0_linux_x86_64.zip
$ wget https://raw.githubusercontent.com/snowplow/factotum/master/samples/echo.factfile
```

These commands will download the 0.6.0 Factotum release, unzip it in your current working directory, and download a sample job for you to run.

Consul is an operational dependency - please see HashiCorp's [getting started guide for Consul][consul-install].

See the [wiki][wiki-home] for further guides and information.

## Developer quickstart

Factotum Server is written in **[Rust][rust-lang]**.

### Using Vagrant

* Clone this repository - `git clone git@github.com:snowplow/factotum-server.git`
* `cd factotum-server`
* Set up a Vagrant box and ssh into it - `vagrant up && vagrant ssh`
   * This will take a few minutes
* `cd /vagrant`
* Compile and run a demo - `cargo run -- --factotum-bin=/vagrant/factotum`
   * Note: You will need to have downloaded the Factotum binary to the path above

### Using stable Rust without Vagrant 

* **[Install Rust][rust-install]**
   * on Linux/Mac - `curl -sSf https://static.rust-lang.org/rustup.sh | sh`
* Clone this repository - `git clone git@github.com:snowplow/factotum-server.git`
* `cd factotum-server`
* Compile and run the server - `cargo run -- --factotum-bin=/vagrant/factotum`
   * Note: You will need to have downloaded the Factotum binary to the path above

## Copyright and license

Factotum Server is copyright 2017-2021 Snowplow Analytics Ltd.

Licensed under the **[Apache License, Version 2.0][license]** (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[license]: http://www.apache.org/licenses/LICENSE-2.0
[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat

[releases]: https://github.com/snowplow/factotum-server/releases
[release-image]: http://img.shields.io/badge/release-0.3.0-6ad7e5.svg?style=flat

[factotum-repo]: https://github.com/snowplow/factotum
[wiki-home]: https://github.com/snowplow/factotum/wiki/Factotum-Server
[user-guide]: https://github.com/snowplow/factotum/wiki/Factotum-Server-User-Guide

[rust-lang]: https://www.rust-lang.org/
[rust-install]: https://www.rust-lang.org/downloads.html
[consul-install]: https://www.consul.io/intro/getting-started/install.html


CHANGELOG
Version 0.3.0 (2021-02-24)
--------------------------
Migrate from Bintray to Github for releases (#12)

Version 0.2.0 (2018-06-26)
--------------------------
Add support for Webhook max-stdouterr-size (#6)
Update Copyright years to 2017-2018 (#7)
Use pip2 instead of pip for OSX build on Travis (#9)

Version 0.1.0 (2017-04-28)
--------------------------
Initial release


src/tests.rs
// Copyright (c) 2017-2021 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0, and
// you may not use this file except in compliance with the Apache License
// Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
// http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License Version 2.0 is distributed on an "AS
// IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the Apache License Version 2.0 for the specific language
// governing permissions and limitations there under.
//

use super::*;

#[test]
fn check_factotum_bin_arg_fail() {
    let expected = Err("Invalid path for Factotum binary at: '/fake/path'".to_string());
    let actual = check_factotum_bin_arg("/fake/path");
    assert_eq!(expected, actual);
}

#[test]
fn check_factotum_bin_arg_success() {
    let expected = Ok(());
    let actual = check_factotum_bin_arg(".");
    assert_eq!(expected, actual);
}

#[test]
fn check_ip_arg_fail() {
    let expected = Err("Invalid IP address: [NOT.AN.IP] - Regex mismatch".to_string());
    let actual = check_ip_arg(&Some("NOT.AN.IP".to_string()));
    assert_eq!(expected, actual);
}

#[test]
fn check_ip_arg_success_with_valid_ip() {
    let expected = Ok(());
    let actual = check_ip_arg(&Some("255.255.255.255".to_string()));
    assert_eq!(expected, actual);
}

#[test]
fn check_ip_arg_success_with_none() {
    let expected = Ok(());
    let actual = check_ip_arg(&None);
    assert_eq!(expected, actual);
}

#[test]
fn is_a_valid_ip_fail() {
    let result = is_a_valid_ip("NOT.AN.IP");
    assert!(result == false);
}

#[test]
fn is_a_valid_ip_success() {
    let result = is_a_valid_ip("255.255.255.255");
    assert!(result == true);
}

#[test]
fn get_log_level_fail() {
    let expected = Err("Unknown log level: 'NOT A LOG LEVEL'\nPlease select a valid log level.".to_string());
    let actual = get_log_level(&Some("NOT A LOG LEVEL".to_string()));
    assert_eq!(expected, actual);
}

#[test]
fn get_log_level_success_valid_level() {
    let expected = Ok(LogLevelFilter::Debug);
    let actual = get_log_level(&Some("DEBUG".to_string()));
    assert_eq!(expected, actual);
}

#[test]
fn get_log_level_success_with_none() {
    let expected = Ok(LogLevelFilter::Warn);
    let actual = get_log_level(&None);
    assert_eq!(expected, actual);
}

src/factotum_server/tests.rs
// Copyright (c) 2017-2021 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0, and
// you may not use this file except in compliance with the Apache License
// Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
// http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License Version 2.0 is distributed on an "AS
// IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the Apache License Version 2.0 for the specific language
// governing permissions and limitations there under.
//

use super::*;
use std::time::Duration;

#[test]
fn worker_manager_spawn_check_queue_and_exit() {
    let (tx, rx) = mpsc::channel();
    let pool = ThreadPool::new(2);
    let persistence = ConsulPersistence::new(None, None, None, None);
    let command_store = commands!["dummy".to_string() => "/tmp/fake_command".to_string()];

    let handle = spawn_worker_manager(tx.clone(), rx, VecDeque::new(), 2, pool.clone(), persistence, command_store);

    let (qtx, qrx) = mpsc::channel();
    let query = Query::new("queue_query", qtx);
    tx.send(Dispatch::CheckQueue(query)).unwrap();

    let output = qrx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert!(output == false);

    tx.send(Dispatch::StopProcessing).unwrap();

    let output = handle.join().unwrap();
    assert_eq!((), output);
}

#[test]
fn send_status_update_success() {
    let (tx, rx) = mpsc::channel();
    let query = Query::new("status_query", tx);
    let pool = ThreadPool::new(2);
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let mut requests_queue = VecDeque::new();
    requests_queue.push_back(job_request);

    send_status_update(query, &mut requests_queue, 10, &pool);

    let actual = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    let expected = DispatcherStatus {
        workers: WorkerStatus {
            total: 2,
            idle: 2,
            active: 0,
        },
        jobs: JobStatus {
            max_queue_size: 10,
            in_queue: 1,
        }
    };
    assert_eq!(expected, actual);
}

#[test]
fn is_queue_full_true() {
    let (tx, rx) = mpsc::channel();
    let query = Query::new("queue_query", tx);
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let mut requests_queue = VecDeque::new();
    requests_queue.push_back(job_request.clone());
    requests_queue.push_back(job_request.clone());

    is_queue_full(query, &mut requests_queue, 2);

    let result = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert!(result == true);
}

#[test]
fn is_queue_full_false() {
    let (tx, rx) = mpsc::channel();
    let query = Query::new("queue_query", tx);
    let mut requests_queue = VecDeque::new();

    is_queue_full(query, &mut requests_queue, 2);

    let result = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert!(result == false);
}

#[test]
fn new_job_request_success_with_threads_available() {
    let (tx, rx) = mpsc::channel();
    let pool = ThreadPool::new(2);
    let persistence = ConsulPersistence::new(None, None, None, None);
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let mut requests_queue = VecDeque::new();

    let result = new_job_request(tx.clone(), &mut requests_queue, &pool, persistence, job_request.clone());
    
    let output = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert_eq!(Dispatch::ProcessRequest, output);
    assert_eq!(Ok(()), result);
    assert!(requests_queue.contains(&job_request));
}

#[test]
fn new_job_request_success_with_no_threads_available() {
    let (tx, rx) = mpsc::channel();
    let pool = ThreadPool::new(1);
    let persistence = ConsulPersistence::new(None, None, None, None);
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let mut requests_queue = VecDeque::new();

    let first = new_job_request(tx.clone(), &mut requests_queue, &pool, persistence.clone(), job_request.clone());
    pool.execute(move || {
        thread::sleep(Duration::from_millis(1000));
    });
    thread::sleep(Duration::from_millis(100));
    let second = new_job_request(tx.clone(), &mut requests_queue, &pool, persistence, job_request.clone());
    
    let output = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert_eq!(Dispatch::ProcessRequest, output);
    assert_eq!(Ok(()), first);
    assert_eq!(Err("No threads available - waiting for a job to complete.".to_string()), second);
    assert!(requests_queue.contains(&job_request));
}

#[test]
fn process_job_request_failure() {
    let (tx, rx) = mpsc::channel();
    let pool = ThreadPool::new(2);
    let persistence = ConsulPersistence::new(None, None, None, None);
    let command_store = commands!["dummy".to_string() => "/tmp/fake_command".to_string()];
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let mut requests_queue = VecDeque::new();
    requests_queue.push_back(job_request.clone());

    process_job_request(tx.clone(), &mut requests_queue, &pool, persistence, command_store);

    let output = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert_eq!(Dispatch::RequestFailure(job_request), output);
}

#[test]
fn complete_job_request_success() {
    let (tx, rx) = mpsc::channel();
    let persistence = ConsulPersistence::new(None, None, None, None);
    let job_request = JobRequest::new("dummy_id_1", "dummy", "/tmp/somewhere", vec![]);

    let outcome = complete_job_request(tx, persistence, job_request);

    let output = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert_eq!(Dispatch::ProcessRequest, output);
    assert_eq!("COMPLETED JOB REQ  jobId:[dummy_id_1]".to_string(), outcome);
}

#[test]
fn failed_job_request_success() {
    let (tx, rx) = mpsc::channel();
    let persistence = ConsulPersistence::new(None, None, None, None);
    let job_request = JobRequest::new("dummy_id_1", "dummy", "/tmp/somewhere", vec![]);

    let outcome = failed_job_request(tx, persistence, job_request);

    let output = rx.recv_timeout(Duration::from_millis(1000)).unwrap();
    assert_eq!(Dispatch::ProcessRequest, output);
    assert_eq!("FAILED JOB REQ jobId:[dummy_id_1]".to_string(), outcome);
}


src/factotum_server/server/tests.rs
// Copyright (c) 2017-2021 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0, and
// you may not use this file except in compliance with the Apache License
// Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
// http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License Version 2.0 is distributed on an "AS
// IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the Apache License Version 2.0 for the specific language
// governing permissions and limitations there under.
//

use super::*;
use std::error::Error;

#[test]
fn create_new_server_manager() {
    let server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://a.webhook.com/".to_string(), true, Some(10_000));

    assert_eq!(server_manager.ip, "0.0.0.0");
    assert_eq!(server_manager.port, 8080);
    assert_eq!(server_manager.state, ::SERVER_STATE_RUN);
    assert_eq!(server_manager.start_time.date(), UTC::today());
    assert_eq!(server_manager.webhook_uri, "http://a.webhook.com/");
    assert!(server_manager.no_colour);
}

#[test]
fn server_manager_is_running() {
    let server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://dummy.test/".to_string(), false, Some(10_000));
    assert!(server_manager.is_running());
}

#[test]
fn server_manager_is_not_running() {
    let mut server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://dummy.test/".to_string(), false, Some(10_000));
    server_manager.state = ::SERVER_STATE_DRAIN.to_string();
    assert_eq!(server_manager.is_running(), false);
}

#[test]
fn server_manager_get_start_time() {
    let server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://dummy.test/".to_string(), false, Some(10_000));
    assert_eq!(server_manager.get_start_time(), UTC::now().format("%F %T %Z").to_string());
}

#[test]
fn server_manager_get_uptime() {
    let server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://dummy.test/".to_string(), false, Some(10_000));
    let uptime = UTC::now().signed_duration_since(server_manager.start_time);
    let seconds = uptime.num_seconds() % 60;
    let minutes = uptime.num_minutes() % 60;
    let hours = uptime.num_hours() % 24;
    let days = uptime.num_days();
    let expected = format!("{} Days, {} Hours, {} Minutes, {} Seconds", days, hours, minutes, seconds);
    assert_eq!(expected, server_manager.get_uptime());
}

#[test]
fn job_request_empty_job_name() {
    let job_request = JobRequest::new("1", "", "/tmp/somewhere", vec![]);
    let command_store = commands![::FACTOTUM.to_string() => "/tmp/fake_path".to_string()];
    let validation_error = JobRequest::validate(job_request.clone(), &command_store).unwrap_err();
    assert_eq!(validation_error, ValidationError::no_output("No valid value found: field 'jobName' cannot be empty".to_string()));
}

#[test]
fn job_request_empty_factfile_path() {
    let job_request = JobRequest::new("1", "dummy", "", vec![]);
    let command_store = commands![::FACTOTUM.to_string() => "/tmp/fake_path".to_string()];
    let validation_error = JobRequest::validate(job_request.clone(), &command_store).unwrap_err();
    assert_eq!(validation_error, ValidationError::no_output("No valid value found: field 'factfilePath' cannot be empty".to_string()));
}

#[test]
fn job_request_invalid_factfile_path() {
    let job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec![]);
    let command_store = commands![::FACTOTUM.to_string() => "/tmp/fake_path".to_string()];
    let validation_error = JobRequest::validate(job_request.clone(), &command_store).unwrap_err();
    assert_eq!(validation_error, ValidationError::no_output("Value does not exist on host for 'factfilePath':'/tmp/somewhere'".to_string()));
}

#[test]
fn job_request_can_append_job_args() {
    let server_manager = ServerManager::new(Some("0.0.0.0".to_string()), 8080, "http://dummy.test/".to_string(), true, Some(10_000));
    let mut job_request = JobRequest::new("1", "dummy", "/tmp/somewhere", vec!["--first-arg".to_string()]);
    JobRequest::append_job_args(&server_manager, &mut job_request);
    assert_eq!(job_request.factfile_args, vec!["--first-arg", "--webhook", "http://dummy.test/", "--max-stdouterr-size", "10000", "--no-colour"]);
}

#[test]
fn settings_request_is_valid() {
    let settings_request = SettingsRequest::new(::SERVER_STATE_RUN);
    let validated_settings_request = SettingsRequest::validate(settings_request.clone()).unwrap();
    assert_eq!(validated_settings_request, settings_request);
}

#[test]
fn settings_request_is_invalid() {
    let settings_request = SettingsRequest::new("NOT A SERVER STATE");
    let validation_error = SettingsRequest::validate(settings_request).err().unwrap();
    assert_eq!(validation_error.description(), "Invalid 'state', must be one of (run|drain)");
}


src/factotum_server/server/mod.rs
// Copyright (c) 2017-2021 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0, and
// you may not use this file except in compliance with the Apache License
// Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
// http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License Version 2.0 is distributed on an "AS
// IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.  See the Apache License Version 2.0 for the specific language
// governing permissions and limitations there under.
//

use std::collections::HashMap;
use std::error;
use std::fmt;
use std::fs::File;
use std::io::Read;
use std::path::Path;
use chrono::{DateTime, UTC};
use crypto::digest::Digest;
use crypto::sha2::Sha256;
use getopts::Options;
use serde_json;

use factotum_server::command::Execution;

#[cfg(test)]
mod tests;

#[derive(Debug)]
pub struct ServerManager {
    pub ip: String,
    pub port: u32,
    pub state: String,
    pub start_time: DateTime<UTC>,
    pub webhook_uri: String,
    pub no_colour: bool,
    pub max_stdouterr_size: Option<usize>,
}

impl ServerManager {
    pub fn new(wrapped_ip: Option<String>, port: u32, webhook_uri: String, no_colour: bool, max_stdouterr_size: Option<usize>) -> ServerManager {
        ServerManager {
            ip: if let Some(ip) = wrapped_ip { ip } else { ::IP_DEFAULT.to_string() },
            port: if port > 0 && port <= 65535 { port } else { ::PORT_DEFAULT },
            state: ::SERVER_STATE_RUN.to_string(),
            start_time: UTC::now(),
            webhook_uri: webhook_uri.to_string(),
            no_colour: no_colour,
            max_stdouterr_size: max_stdouterr_size,
        }
    }

    pub fn is_running(&self) -> bool {
        self.state == ::SERVER_STATE_RUN
    }

    pub fn get_start_time(&self) -> String {
        self.start_time.format("%F %T %Z").to_string()
    }

    pub fn get_uptime(&self) -> String {
        let uptime = UTC::now().signed_duration_since(self.start_time);
        let seconds = uptime.num_seconds() % 60;
        let minutes = uptime.num_minutes() % 60;
        let hours = uptime.num_hours() % 24;
        let days = uptime.num_days();
        format!("{} Days, {} Hours, {} Minutes, {} Seconds", days, hours, minutes, seconds)
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobRequest {
    #[serde(default)]
    pub job_id: String,
    pub job_name: String,
    pub factfile_path: String,
    pub factfile_args: Vec<String>
}

impl JobRequest {
    #[allow(dead_code)]
    pub fn new(job_id: &str, job_name: &str, factfile_path: &str, factfile_args: Vec<String>) -> JobRequest {
        JobRequest {
            job_id: job_id.to_owned(),
            job_name: job_name.to_owned(),
            factfile_path: factfile_path.to_owned(),
            factfile_args: factfile_args,
        }
    }

    pub fn validate<U: Execution>(request: JobRequest, command_store: &U) -> Result<JobRequest, ValidationError> {
        // check job name not empty
        // check factfile path not empty
        // check factfile args not empty
        if request.job_name == "" {
            let message = format!("No valid value found: field 'jobName' cannot be empty");
            error!("{}", message);
            return Err(ValidationError::no_output(message))
        } else if request.factfile_path == "" {
            let message = format!("No valid value found: field 'factfilePath' cannot be empty");
            error!("{}", message);
            return Err(ValidationError::no_output(message))
        }
        // check valid factfile path exists
        if !Path::new(&request.factfile_path).exists() {
            let message = format!("Value does not exist on host for 'factfilePath':'{}'", request.factfile_path);
            error!("{}", message);
            return Err(ValidationError::no_output(message))
        }
        // attempt dry run
        let cmd_path = try!(command_store.get_command(::FACTOTUM));
        let mut cmd_args = vec!["run".to_string(), request.factfile_path.clone(), "--dry-run".to_string()];
        cmd_args.extend_from_slice(request.factfile_args.as_slice());
        match command_store.execute(cmd_path, cmd_args) {
            Ok(_) => {
                debug!("Dry run success");
            },
            Err(e) => {
                error!("{}", e);
                return Err(ValidationError::no_output(e))
            }
        }
        // generate unique job id
        let mut request = request;
        let tags = match extract_tags(&request.factfile_args) {
            Ok(extracted) => extracted,
            Err(e) => {
                error!("{}", e);
                return Err(ValidationError::no_output(e))
            }
        };
        request.job_id = match generate_id(&request.factfile_path, tags) {
            Ok(id) => id,
            Err(e) => {
                error!("{}", e);
                return Err(ValidationError::no_output(e))
            }
        };
        Ok(request)
    }

    pub fn append_job_args(server: &ServerManager, job: &mut JobRequest) {
        if server.webhook_uri != "" {
            job.factfile_args.push("--webhook".to_string());
            job.factfile_args.push(server.webhook_uri.clone());

            // Only required with webhook
            if let Some(max_bytes) = server.max_stdouterr_size.clone() {
                job.factfile_args.push("--max-stdouterr-size".to_string());
                job.factfile_args.push(max_bytes.to_string());
            };
        }
        if server.no_colour {
            job.factfile_args.push("--no-colour".to_string());
        }
    }
}

impl PartialEq for JobRequest {
    fn eq(&self, other: &JobRequest) -> bool {
        self.job_id        == other.job_id &&
        self.job_name      == other.job_name &&
        self.factfile_path == other.factfile_path &&
        self.factfile_args == other.factfile_args
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SettingsRequest {
    pub state: String
}

impl PartialEq for SettingsRequest {
    fn eq(&self, other: &SettingsRequest) -> bool {
        self.state == other.state
    }
}

impl SettingsRequest {
    #[allow(dead_code)]
    pub fn new(state: &str) -> SettingsRequest {
        SettingsRequest {
            state: state.to_owned()
        }
    }

    pub fn validate(request: SettingsRequest) -> Result<SettingsRequest, ValidationError> {
        match request.state.as_ref() {
            ::SERVER_STATE_RUN | ::SERVER_STATE_DRAIN => Ok(request),
            _ => Err(ValidationError::no_output(format!("Invalid 'state', must be one of ({}|{})", ::SERVER_STATE_RUN, ::SERVER_STATE_DRAIN)))
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct ValidationError {
    pub error: String,
    pub stdout: String,
    pub stderr: String
}

impl ValidationError {
    pub fn new(error: String, stdout: String, stderr: String) -> ValidationError {
        ValidationError {
            error: error,
            stdout: stdout,
            stderr: stderr,
        }
    }

    pub fn no_output(error: String) -> ValidationError {
        ValidationError::new(error, String::new(), String::new())
    }
}

impl fmt::Display for ValidationError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Validation Error: {}", self.error)
    }
}

impl error::Error for ValidationError {
    fn description(&self) -> &str {
        &self.error
    }
}

impl From<String> for ValidationError {
    fn from(err: String) -> ValidationError {
        ValidationError::no_output(err)
    }
}

fn generate_id(factfile: &str, tags: Option<HashMap<String, String>>) -> Result<String, String> {
    let mut fh = try!(File::open(factfile)
        .map_err(|e| format!("Could not open '{}' for reading: {}", factfile, e)));
    let mut file = String::new();
    try!(fh.read_to_string(&mut file).map_err(|e| format!("Could not read '{}': {}", factfile, e)));
    let schema: serde_json::Value = try!(serde_json::from_str(&file).map_err(|e| e.to_string()));
    let ff: String = try!(serde_json::to_string(&schema).map_err(|e| e.to_string()));
    let mut job_digest = Sha256::new();
    job_digest.input_str(&ff);

    if let Some(ref tags_map) = tags {
        let mut sorted_keys:Vec<_> = tags_map.keys().collect();
        sorted_keys.sort();
        for key in sorted_keys {
            job_digest.input_str(key);
            job_digest.input_str(&tags_map[key]);
        }
    }

    Ok(job_digest.result_str())
}

fn extract_tags(factfile_args: &Vec<String>) -> Result<Option<HashMap<String, String>>, String> {
    let mut opts = Options::new();
    opts.optmulti("", "tag", "Add Factotum job metadata (tags).", "TAG");
    opts.optmulti("", "constraint", "Checks for an external constraint that will prevent execution; allowed constraints (host).", "CONSTRAINT");
    opts.optopt("", "start", "Begin at specified task.", "TASK");
    opts.optopt("", "env", "Supply JSON to define mustache variables in Factfile.", "ENV");
    opts.optopt("", "webhook", "Post updates on job execution to the specified URL.", "URL");
    opts.optopt("", "output", "File to print output to. Used with `dot`.", "FILE");
    opts.optflag("", "dry-run", "Pretend to execute a Factfile, showing the commands that would be executed. Can be used with other options.");
    opts.optflag("", "no-colour", "Turn off ANSI terminal colours/formatting in output.");
    opts.optflag("", "overwrite", "Overwrite the output file if it exists.");
    opts.optopt("", "max-stdouterr-size", "The maximum size of the individual stdout/err sent via the webhook functions for job updates.", "BYTES");

    let matches = match opts.parse(factfile_args) {
        Ok(m) => m,
        Err(e) => {
            return Err(format!("Error parsing tags in factfile args: {:?}", e))
        },
    };

    if matches.opt_present("tag") {
        let tag_map = get_tag_map(&matches.opt_strs("tag"));
        trace!("Extracted tags: {:?}", tag_map);
        Ok(Some(tag_map))
    } else {
        trace!("NO TAGS FOUND");
        Ok(None)
    }
}

fn get_tag_map(args: &Vec<String>) -> HashMap<String, String> {
    let mut arg_map: HashMap<String, String> = HashMap::new();

    for arg in args.iter() {
        let split = arg.split(",").collect::<Vec<&str>>();
        if split.len() >= 2 && split[0].trim().is_empty() == false {
            let key = split[0].trim().to_string();
            let value = split[1..].join("").trim().to_string();
            arg_map.insert(key, value);
        } else if split.len() == 1 && split[0].trim().is_empty() == false {
            let key = split[0].trim().to_string();
            let value = "".to_string();
            arg_map.insert(key, value);
        }
    }

    arg_map
}
